tools/src/main/java/com/nedap/archie/creation/OpenEhrRmInstanceGenerator.java

```diff
@@ -0,0 +1,378 @@
+        Map<String, Object> archetypeDetails = new LinkedHashMap<>();
+        archetypeDetails.put(typePropertyName, "ARCHETYPED");
+        Map<String, Object> archetypeId = new LinkedHashMap<>();
+        archetypeId.put(typePropertyName, "ARCHETYPE_ID");
+        archetypeId.put("value", archetypeIdValue);
+        archetypeDetails.put("archetype_id", archetypeId); //TODO: add template id?
+        archetypeDetails.put("rm_version", "1.0.4");
+        return archetypeDetails;
+    }
+
+    public void addAdditionalPropertiesAtEnd(BmmClass classDefinition, Map<String, Object> result, CObject cObject) {
+        String rmTypeName = classDefinition.getType().getBaseClass().getName();
+        switch (rmTypeName.toUpperCase()) {
+            case "DV_CODED_TEXT":
+                fixCodedText(result, cObject);
+                break;
+            case "CODE_PHRASE":
+                fixCodePhrase(result, cObject);
+                break;
+            case "HISTORY":
+                fixHistory(result);
+                break;
+            case "ELEMENT":
+                fixElement(result);
+                break;
+            case "DV_MULTIMEDIA":
+                fixDvMultimedia(result);
+                break;
+            case "PARTY_RELATED":
+                FixPartyRelated(result);
+                break;
+            case "DV_PROPORTION":
+                fixDvProportion(result);
+                break;
+            case "DV_URI":
+            case "DV_EHR_URI":
+                fixDvUri(result);
+                break;
+            case "PARTY_REF":
+                fixPartyRef(result);
+                break;
+            case "DV_INTERVAL":
+                fixDvInterval(result);
+                break;
+        }
+    }
+
+    private void fixDvInterval(Map<String, Object> result) {
+        Boolean lowerUnbounded = (Boolean) result.get("lower_unbounded");
+        Boolean lowerIncluded = (Boolean) result.get("lower_included");
+        Boolean upperUnbounded = (Boolean) result.get("upper_unbounded");
+        Boolean upperIncluded = (Boolean) result.get("upper_included");
+
+        if(result.get("lower") == null) {
+            result.put("lower_unbounded", true);
+        } else {
+            result.put("lower_unbounded", false);
+        }
+        if(result.get("upper") == null) {
+            result.put("upper_unbounded", true);
+        } else {
+            result.put("upper_unbounded", false);
+        }
+        //on partially unconstrained intervals, default values will have been added that are not valid. Correct them here.
+        if(lowerUnbounded != null && lowerIncluded != null) {
+            if(lowerUnbounded) {
+                result.put("lower_included", false);
+            }
+        }
+        if(upperUnbounded != null && upperIncluded != null) {
+            if(upperUnbounded) {
+                result.put("upper_included", false);
+            }
+        }
```

Comment:
You're using the old values for `xUnbounded` on lines 203/204 and 208/209. Maybe this is better?

```suggestion
        Boolean lowerIncluded = (Boolean) result.get("lower_included");
        Boolean upperIncluded = (Boolean) result.get("upper_included");

        if(result.get("lower") == null) {
            result.put("lower_unbounded", true);
            if(lowerIncluded != null) {
                result.put("lower_included", false);
            }
        } else {
            result.put("lower_unbounded", false);
        }
        if(result.get("upper") == null) {
            result.put("upper_unbounded", true);
            if(upperIncluded != null) {
                result.put("upper_included", false);
            }
        } else {
            result.put("upper_unbounded", false);
        }
```

Reply:
better indeed. Added it, and credited you in the commit message.